<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Вывод-первых-строк-и-изучение-полученной-информации" data-toc-modified-id="Вывод-первых-строк-и-изучение-полученной-информации-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Вывод первых строк и изучение полученной информации</a></span></li><li><span><a href="#Замена-типов-данных" data-toc-modified-id="Замена-типов-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Замена типов данных</a></span></li><li><span><a href="#Обработка-явных-дубликатов" data-toc-modified-id="Обработка-явных-дубликатов-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Обработка явных дубликатов</a></span></li><li><span><a href="#Промежуточный-вывод" data-toc-modified-id="Промежуточный-вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Промежуточный вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Дано" data-toc-modified-id="Дано-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Дано</a></span></li><li><span><a href="#Доказательство" data-toc-modified-id="Доказательство-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Доказательство</a></span></li><li><span><a href="#Промежуточный-вывод" data-toc-modified-id="Промежуточный-вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Промежуточный вывод</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Создадим-функцию-для-получения-случайной-квадратной-и-закодированной-матрицы" data-toc-modified-id="Создадим-функцию-для-получения-случайной-квадратной-и-закодированной-матрицы-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Создадим функцию для получения случайной квадратной и закодированной матрицы</a></span></li><li><span><a href="#Обучим-и-проверим-модель" data-toc-modified-id="Обучим-и-проверим-модель-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Обучим и проверим модель</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

**Описание данных**

Набор данных находится в файле /datasets/insurance.csv.

- Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.
- Целевой признак: количество страховых выплат клиенту за последние 5 лет.

## Загрузка данных

### Импорт библиотек

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

### Вывод первых строк и изучение полученной информации

In [2]:
def my_func(x):
    print('Исходный датафрейм')
    display(x.head(20))
    print('Общая информация о датафрейме')
    display(x.info())
    print('Количество пустых значений в датафрейме')
    display(x.isna().sum())
    print('Количество явных дубликатов в датафрейме')
    display(x.duplicated().sum())
    print('Описательная статистика')
    display(x.describe())
    print('Зависимость признаков')
    cmap = cmap=sns.diverging_palette(5, 250, as_cmap=True)
    def magnify():
        return [dict(selector="th",
                     props=[("font-size", "7pt")]),
                dict(selector="td",
                     props=[('padding', "0em 0em")]),
                dict(selector="th:hover",
                     props=[("font-size", "12pt")]),
                dict(selector="tr:hover td:hover",
                     props=[('max-width', '200px'),
                            ('font-size', '12pt')])
    ]
    corr = round(x.corr(), 2)
    display(corr.style.background_gradient(cmap, axis=1)\
        .set_properties(**{'max-width': '80px', 'font-size': '10pt'})\
        .set_caption("Hover to magify")\
        .set_precision(2)\
        .set_table_styles(magnify()))

In [3]:
try:
    data = pd.read_csv('/datasets/insurance.csv')
except:
    data = pd.read_csv('C:/Users/daria/Датасеты для проектов/10 Линейная алгебра - защита персональных данных клиентов/insurance.csv')
my_func(data)

Исходный датафрейм


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0     1     41.0   49600.0            1                  0
1     0     46.0   38000.0            1                  1
2     0     29.0   21000.0            0                  0
3     0     21.0   41700.0            2                  0
4     1     28.0   26100.0            0                  0
5     1     43.0   41000.0            2                  1
6     1     39.0   39700.0            2                  0
7     1     25.0   38600.0            4                  0
8     1     36.0   49700.0            1                  0
9     1     32.0   51700.0            1                  0
10    1     25.0   36600.0            1                  0
11    1     38.0   29300.0            0                  0
12    0     23.0   39500.0            3                  0
13    0     21.0   55000.0            0                  0
14    0     40.0   43700.0            1                  0
15    1     34.0   23300.0            0                  0
16    1     26.0   48900.0            2                  0
17    1     41.0   33200.0            2                  0
18    1     42.0   49700.0            0                  1
19    1     27.0   36900.0            0                  0

Общая информация о датафрейме
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

Количество пустых значений в датафрейме


Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Количество явных дубликатов в датафрейме


153

Описательная статистика


Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

Зависимость признаков


C:\Users\daria\AppData\Local\Temp\ipykernel_9736\2258407746.py:26: FutureWarning: this method is deprecated in favour of `Styler.format(precision=..)`
  display(corr.style.background_gradient(cmap, axis=1)\


### Замена типов данных

In [4]:
to_convert = {'Возраст', 'Зарплата'}
for f in to_convert:
        data[f] = data[f].astype('int')

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                5000 non-null   int64
 1   Возраст            5000 non-null   int32
 2   Зарплата           5000 non-null   int32
 3   Члены семьи        5000 non-null   int64
 4   Страховые выплаты  5000 non-null   int64
dtypes: int32(2), int64(3)
memory usage: 156.4 KB


### Обработка явных дубликатов

In [6]:
data.duplicated().sum()

153

In [7]:
data[data.duplicated()]

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281     1       39     48100            1                  0
488     1       24     32900            1                  0
513     0       31     37400            2                  0
718     1       22     32600            1                  0
785     0       20     35800            0                  0
...   ...      ...       ...          ...                ...
4793    1       24     37800            0                  0
4902    1       35     38700            1                  0
4935    1       19     32700            0                  0
4945    1       21     45800            0                  0
4965    0       22     40100            1                  0

[153 rows x 5 columns]

In [8]:
data[(data['Возраст'] == 21) & (data['Зарплата'] == 41700)]

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
3       0       21     41700            2                  0
3894    0       21     41700            2                  0

In [9]:
data = data.drop_duplicates()

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                4847 non-null   int64
 1   Возраст            4847 non-null   int32
 2   Зарплата           4847 non-null   int32
 3   Члены семьи        4847 non-null   int64
 4   Страховые выплаты  4847 non-null   int64
dtypes: int32(2), int64(3)
memory usage: 189.3 KB


### Промежуточный вывод

В данном исследовании были выявлены следующие характеристики:

1. Изначальное количество строк в исходном датафрейме составляет 5000.
2. Пропущенные значения отсутствуют.
3. Количество явных дубликатов 153. Была проведена работа по устранению дубликатов, количество строк в финальном датафрейме - 4847.
4. Тип данных в столбцых "Возраст" и "Зарплата" был заменен с float64 на целочисленный.
5. Присутствует достаточно сильная корреляция между страховыми выплатами и возрастом - 0.65. Остальные признаки не зависят друг от друга.

## Умножение матриц

### Дано

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

### Доказательство

**Ответ:** Результаты моделей на исходных и преобразованных признаках будут одинаковыми.

**Обоснование:** Формула предсказания линейной регрессии a = Xw, для модели на преобразованных признаках a' = X'w' = (XP)w', где P - случайная обратимая матрица. Нужно доказать, что a = a'. Для этого берем **формулу обучения** $w = (X^T X)^{-1} X^T y$, записываем её для w', подставляя XP вместо X. Выполняем для неё преобразование так, чтобы получить формулу w' от w. Затем подставляем найденное w' в формулу для a' и производим сокращения и упрощения. Получим a' = a. Матрица признаков X в общем случае прямоугольная, и значит для неё не существует обратной матрицы.

Свойства транспонированной матрицы:
$$
(A \cdot B)^T = B^T \cdot A^T
$$
Свойства обратных матриц:
$$
(A \cdot B)^{-1} = A^{-1} \cdot B^{-1}
$$
При умножении матрицы на обратную ей тоже получается единичная матрица:
$$
(A \cdot A)^{-1} = E
$$

Доказательство:
$$
w' = ((XP)^T \cdot XP)^{-1} (XP)^Ty = (P^TX^T \cdot XP)^{-1} P^TX^Ty = (P^T(X^TX)P)^{-1} P^TX^Ty = ((X^TX)P)^{-1}(P^T)^{-1} P^TX^Ty = P^{-1}(X^TX)^{-1} (P^T)^{-1}P^TX^Ty = P^{-1}(X^TX)^{-1} EX^Ty = P^{-1}(X^TX)^{-1} X^Ty
$$

Напомним, что формула обучения выглядит так:

$$
w = (X^T X)^{-1} X^T y
$$

В ходе доказательства получили формулу w' от w:

$$
w' = P^{-1} (X^T X)^{-1} X^T y
$$

Чтд

### Промежуточный вывод

При кодировании матрицы признаков X матрицей P формула обучения приобретает вид: 
$w' = P^{-1} (X^T X)^{-1} X^T y$, при условии, что матрица P должна быть квадратной. Отсюда $a' = X'w' = XPP^{-1}w = a$. В результате качество моделей на исходных и преобразованных признаках будет одинаковым.

Ответ на задание: "Укажите, как связаны параметры линейной регрессии в исходной задаче и в преобразованной". Веса исходных и преобразованных признаков связаны через обратную матрицу шифрования $P^{-1}$:

$$
w' = P^{-1} (X^T X)^{-1} X^T y
$$
$$
w' = P^{-1} w
$$
        
или 
$$
\frac{w}{w'} = \frac{(X^T X)^{-1} X^T y}{P^{-1} (X^T X)^{-1} X^T y} = \frac{1}{P^{-1}}
$$
        
        
$$
w' = P^{-1} w
$$

## Алгоритм преобразования

**Алгоритм**

Первый шаг:
- Произведем разделение данных на целевой признак - target и остальные признаки - features.
- Сгенерируем случайную квадратную матрицу P размера nхn для матрицы X размера mхm и проверим на обратимость. В случае, если она необратима, то необходимо сгенерировать новые матрицы до момента, когда получится обратимая матрица.
- Умножим матрицу P на выборку с остальными признаками (features) и получим закодированную матрицу (data_coded).

Второй шаг:

- Создадим расширенную матрицу со столбцом из единиц.
- Используем формулу $w = (X^T X)^{-1} X^T y$.

Третий шаг:
- Обучим модели методом линейной регрессии и проверим - совпали ли метрики R2 на исходной матрице и закодированной.

**Обоснование**

По причине того, что предсказания a = a', что было доказано выше, качество линейной регрессии не изменится.

## Проверка алгоритма

### Создадим функцию для получения случайной квадратной и закодированной матрицы

In [11]:
features = data.drop('Страховые выплаты', axis=1)
target = data['Страховые выплаты']

def get_rand_matrix(data):
    det = 0
    while det == 0:
        matrix = np.random.normal(size=(data.shape[1], data.shape[1]))
        det = np.linalg.det(matrix)
    return matrix

matrix = get_rand_matrix(features)
matrix

data_coded = features @ matrix

display('Закодированная матрица')
display(data_coded)

'Закодированная матрица'

0            1             2            3
0    -2235.276324  8921.662909 -69272.294616  7719.404115
1    -1711.005914  6843.794354 -53079.184795  5932.230727
2     -945.415051  3783.717065 -29334.833966  3282.350288
3    -1882.102436  7493.823315 -58235.163390  6475.734564
4    -1174.713303  4697.954072 -36453.409421  4068.687408
...           ...          ...           ...          ...
4995 -1609.223027  6421.618508 -49860.019899  5555.742231
4996 -2365.049546  9419.867220 -73181.487911  8145.808778
4997 -1529.183412  6094.104684 -47343.355344  5267.968663
4998 -1471.829133  5880.826889 -45666.002051  5083.532549
4999 -1830.201653  7299.738770 -56699.989734  6312.016319

[4847 rows x 4 columns]

### Обучим и проверим модель

In [12]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = (np.linalg.inv(X.T.dot(X)).dot(X.T)).dot(y)
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0

Обучение модели методом линейной регрессии на исходной матрице: 

In [13]:
def func_lr(features, target):
    
    features_train, features_test, target_train, target_test = train_test_split(features,
                                                                                target,
                                                                                test_size=0.2,
                                                                                random_state=12345)
    model = LinearRegression()
    model.fit(features_train, target_train)

    predictions_train = model.predict(features_train)
    predictions_test = model.predict(features_test)

    display(r2_score(target_train, predictions_train))
    display(r2_score(target_test, predictions_test))

In [14]:
func_lr(features, target)

0.433348163455491

0.41605492298279634

Обучение модели методом линейной регрессии на закодированной матрице: 

In [15]:
func_lr(data_coded, target)

0.433348163455497

0.41605492004052125

## Общий вывод

В данном исследовании были выявлены следующие характеристики:

Изначальное количество строк в исходном датафрейме составляет 5000.
Пропущенные значения отсутствуют.
Количество явных дубликатов 153. Была проведена работа по устранению дубликатов, количество строк в финальном датафрейме - 4847.
Тип данных в столбцых "Возраст" и "Зарплата" был заменен с float64 на целочисленный.
Присутствует достаточно сильная корреляция между страховыми выплатами и возрастом - 0.65. Остальные признаки не зависят друг от друга.

При кодировании матрицы признаков X матрицей P формула обучения приобретает вид: 
$w' = P^{-1} (X^T X)^{-1} X^T y$, при условии, что матрица P должна быть квадратной. Отсюда $a' = X'w' = XPP^{-1}w = a$. Было доказано, что в результате преобразования качество моделей на исходных и преобразованных признаках будет одинаковым.

Далее создали алгоритм преобразования данных для решения задачи и выполнили проверку алгоритма, которая показала одинаковое значение R2 на обеих выборках: для обучения 0.430201, на тестовой 0.41605.

Итог - данные клиентов страховой компании «Хоть потоп» защищены. 